Adding perf tests for the new System.Math/MathF APIs added in netcoreapp2.1/3.0

tannergooding opened the pull request:
This adds perf tests for the `Acosh`, `Asinh`, `Atanh`, and `Cbrt` functions added to `System.Math` and `System.MathF` in netcoreapp2.1 and the `FusedMultiplyAdd`, `ILogB`, `Log2`, and `ScaleB` functions added to `System.Math` and `System.MathF` in netcoreapp3.0

tannergooding commented:
There are some more tests I wanted to add for new tests in netcoreapp3.0; but it looks like the consumed CoreCLR/CoreFX bits aren't quite in sync....

I would also like to fixup the existing tests that were ported, but @adamsitnik indicated that might need approval first.

tannergooding commented on src/benchmarks/micro/ValuesGenerator.cs:
```diff
@@ -87,12 +94,34 @@ private static T GenerateValue<T>(Random random)
             throw new NotImplementedException($"{typeof(T).Name} is not implemented");
         }
 
+        private static T GenerateValue<T>(Random random, T minValue, T maxValue)
```
Many of the `System.Math` functions have special handling in the JIT to either cache computed results or to perform constant folding. So we want to generate a unique input per iteration that is within the allowed domain.

jorive replied:
@tannergooding Could you please add this as a function header/comment?

tannergooding replied:
Fixed.

jorive commented on src/benchmarks/micro/coreclr/Math/Functions/Double/AcoshDouble.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Functions
+{
+    [BenchmarkCategory(Categories.CoreCLR)]
+    public class AcoshDouble
```
The naming convention here should be:
class name: `public partial class MathTests`
benchmark method name: `AcoshDoubleBenchmark`

for reference please take a look at [AcosDouble.cs](https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/coreclr/Math/Functions/Double/AcosDouble.cs)

tannergooding replied:
How would I modify `IterationSetup` to only apply to a given test (given that we need a different one for each `[Benchmark]` to generate the input value from the given domain)?

AndreyAkinshin replied:
@tannergooding Setup/Clean attributes have `Target` property which allows specifying target benchmark for each Setup/Cleanup methods. Thus, you can write separate IterationSetup methods for each benchmark (a usage example: https://benchmarkdotnet.org/articles/samples/IntroSetupCleanupTarget.html)

brianrob commented:
cc: @billwert

tannergooding commented:
Currently getting:
```
// * Warnings *
MinIterationTime
  MathTests.AcoshDoubleTest: InvocationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, UnrollFactor=1, WarmupCount=1 -> MinIterationTime = 400.0000 ns which is very small. It's recommended to increase it.
  MathTests.AsinhDoubleTest: InvocationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, UnrollFactor=1, WarmupCount=1 -> MinIterationTime = 500.0000 ns which is very small. It's recommended to increase it.
  MathTests.AtanhDoubleTest: InvocationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, UnrollFactor=1, WarmupCount=1 -> MinIterationTime = 700.0000 ns which is very small. It's recommended to increase it.
  MathTests.AtanhSingleTest: InvocationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, UnrollFactor=1, WarmupCount=1 -> MinIterationTime = 800.0000 ns which is very small. It's recommended to increase it.
MultimodalDistribution
  MathTests.AcoshDoubleTest: InvocationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, UnrollFactor=1, WarmupCount=1 -> It seems that the distribution is multimodal (mValue = 4.25)
  MathTests.AtanhDoubleTest: InvocationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, UnrollFactor=1, WarmupCount=1 -> It seems that the distribution is bimodal (mValue = 3.75)
  MathTests.CbrtDoubleTest: InvocationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, UnrollFactor=1, WarmupCount=1  -> It seems that the distribution is bimodal (mValue = 3.56)
  MathTests.AcoshSingleTest: InvocationCount=1, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, UnrollFactor=1, WarmupCount=1 -> It seems that the distribution can have several modes (mValue = 3)
  MathTests.TanSingleBenchmark: IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1                                 -> It seems that the distribution can have several modes (mValue = 3.14)
```

What is the recommended "MinIterationTime" and how can I determine the correct value to scale these by?
* @adamsitnik mentioned that I can use `OperationsPerInvoke` and invoke the call multiple times in the benchmark body; so I just need to know how to determine how many times I should be invoking the call per iteration

adamsitnik commented on src/benchmarks/micro/ValuesGenerator.cs:
```diff
@@ -87,12 +94,40 @@ private static T GenerateValue<T>(Random random)
+            if (typeof(T) == typeof(int))
+                return (T)(object)random.Next((int)(object)minValue, (int)(object)maxValue);
+            if (typeof(T) == typeof(double))
+                return (T)(object)BitConverter.Int64BitsToDouble(
```
Maybe we could use a simpler approach? From [SO](https://stackoverflow.com/a/1064907/5852046)

`random.NextDouble() * (maximum - minimum) + minimum;`

tannergooding replied:
I believe the current approach gives the best distribution of results, over the desired range, without falling into a pit of failure due to IEEE rounding and result representation issues.

adamsitnik replied:
I am sure it is. However, to understand this code one needs an expert knowledge. You have it, others (including me) don't ;)

Could you add a comment explaining how it works and some unit tests to make sure it's correct?

adamsitnik commented on src/benchmarks/micro/ValuesGenerator.cs:
```diff
@@ -87,12 +94,40 @@ private static T GenerateValue<T>(Random random)
+                );
+            if (typeof(T) == typeof(float))
+            {
+                int result = random.Next(Unsafe.As<T, int>(ref minValue), Unsafe.As<T, int>(ref maxValue));
```
maybe we could take the double result and cast it to float?  `(T)(object)(float)GenerateValue<double>()`

tannergooding replied:
Same comment as above.

adamsitnik commented on src/benchmarks/micro/ValuesGenerator.cs:
```diff
@@ -21,6 +22,12 @@ public static T GetNonDefaultValue<T>()
 
+        public static T GetValue<T>(T minValue, T maxValue)
+        {
+            var random = new Random(Seed);
```
this is very tricky: this method will always return the same result because we use a const seed.

```cs
for (int i = 0; i < 10; i++)
    Console.WriteLine(ValuesGenerator.GetValue(1.0, double.PositiveInfinity));
```

```
3,98113421984948E+20
3,98113421984948E+20
3,98113421984948E+20
3,98113421984948E+20
3,98113421984948E+20
3,98113421984948E+20
3,98113421984948E+20
3,98113421984948E+20
3,98113421984948E+20
3,98113421984948E+20
```

In general, this is the desired approach. However, in our case, we want to have a different number per iteration to make sure the JIT does not cache the result?

tannergooding replied:
Ah, right. We would need to have it keep the same random that was initially seeded, which would also require the tests to run in a deterministic order.

adamsitnik replied:
@gfoidl we should rather create it inside a method and pass to other methods like here:

https://github.com/dotnet/performance/blob/c4efbbb55e0a5d8f4408c9806bdb51b300a261fc/src/benchmarks/micro/ValuesGenerator.cs#L29-L35

so anytime the method is invoked it produces the same results

danmosemsft replied:
> JIT does not cache the result?

By what mechanism would the JIT do that? It doesn't know enough to know it's the same result each time does it?

tannergooding replied:
A number of the `System.Math` functions are intrinsicly recognized by the JIT and will under go constant folding since they should be deterministic.

tannergooding commented:
> I can see that the time it takes to execute most of the tested methods is dependent on the input. For example time(Math.Acosh(1.0)) != time(Math.Acosh(100.0)).

Yes, many of these methods special-case certain inputs or ranges of inputs for speed purposes.

billwert commented:
/azp run

tannergooding commented:
I'll try to get around to rebasing this and resolving remaining feedback sometime today.

billwert commented:
@tannergooding no worries. I was wondering if I could trick it into running anyway. :)

jorive commented:
@tannergooding Could you please follow the naming convention in the recently renamed Math benchmarks here https://github.com/dotnet/performance/tree/master/src/benchmarks/micro/coreclr/Math/Functions ?

tannergooding commented:
I've fixed up the new tests to match the existing namespace/layout and reverted the behavior back to be closer to what I had in https://github.com/dotnet/coreclr/pull/21289.

The reasoning being that they not only more closely mirror the behavior of the other Math/MathF tests, but it also ensures that we aren't falling subject to issues with the cost of the functions being too small or to the JIT performing constant folding on them and causing incorrect numbers.

tannergooding commented:
I think, if we want to fix up the tests to do some different behavior, we should instead do that all at once and in a separate PR.

tannergooding commented:
I think this is ready for a final review pass.

billwert commented:
LGTM. @tannergooding agreed to open an issue to capture recommendations for what the future of this entire suite looks like.

tannergooding commented:
I logged https://github.com/dotnet/performance/issues/423 which captures the general discussion had above and some of the issues to look out for.
